packages/runtime/src/runtime.ts

```diff
@@ -736,144 +587,29 @@ export class Runtime {
-    for (const [index, txnParam] of txnParameters.entries()) {
-      this.assertAmbiguousTxnParams(txnParam);
-      if (txnParam.sign === SignType.LogicSignature) {
-        this.ctx.tx = this.ctx.gtxs[index]; // update current tx to index of stateless
-        this.validateLsigAndRun(txnParam);
-        this.ctx.tx = this.ctx.gtxs[0]; // after executing stateless tx updating current tx to default (index 0)
-      }
-
-      // https://developer.algorand.org/docs/features/asc1/stateful/#the-lifecycle-of-a-stateful-smart-contract
-      switch (txnParam.type) {
-        case TransactionType.CallNoOpSSC:
-        case TransactionType.CloseSSC:
-        case TransactionType.DeleteSSC: {
-          const appParams = this.getApp(txnParam.appId);
-          this.run(appParams["approval-program"], ExecutionMode.STATEFUL);
-          break;
-        }
-        case TransactionType.ClearSSC: {
-          const appParams = this.assertAppDefined(txnParam.appId, this.getApp(txnParam.appId));
-          const fromAccount = this.getAccount(txnParam.fromAccount.addr);
-          try {
-            this.run(appParams["clear-state-program"], ExecutionMode.STATEFUL);
-          } catch (error) {
-            // if transaction type is Clear Call, remove the app first before throwing error (rejecting tx)
-            // https://developer.algorand.org/docs/features/asc1/stateful/#the-lifecycle-of-a-stateful-smart-contract
-            fromAccount.closeApp(txnParam.appId); // remove app from local state
-            throw error;
-          }
-
-          fromAccount.closeApp(txnParam.appId); // remove app from local state
-          break;
-        }
-        case TransactionType.FreezeAsset: {
-          const asset = this.getAssetDef(txnParam.assetID);
-          if (asset.freeze !== txnParam.fromAccount.addr) {
-            throw new RuntimeError(RUNTIME_ERRORS.ASA.FREEZE_ERROR, { address: asset.freeze });
-          }
-          break;
-        }
-        case TransactionType.RevokeAsset: {
-          const asset = this.getAssetDef(txnParam.assetID);
-          if (asset.clawback !== txnParam.fromAccount.addr) {
-            throw new RuntimeError(RUNTIME_ERRORS.ASA.CLAWBACK_ERROR, { address: asset.clawback });
-          }
-          break;
-        }
-        case TransactionType.DestroyAsset:
-        case TransactionType.ModifyAsset: {
-          const asset = this.getAssetDef(txnParam.assetID);
-          if (asset.manager !== txnParam.fromAccount.addr) {
-            throw new RuntimeError(RUNTIME_ERRORS.ASA.MANAGER_ERROR, { address: asset.manager });
-          }
-          break;
-        }
-      }
-    }
+    // Run TEAL program associated with each transaction and
+    // then execute the transaction without interacting with store.
+    this.ctx.processTransactions(txnParameters);
 
     // update store only if all the transactions are passed
-    this.updateFinalState(txnParameters);
+    this.store = this.ctx.state;
   }
```

Comment:
just to confirm, i don't see the update  `this.ctx = {..}` to `this.ctx = new Ctx(..)` for app/asset creation & opt-in. Is that fine the previous way?

Reply:
i think it is fine, otherwise test should fail.
-  asset creation, opt-in doesn't require ctx update.
- app creation, update and optin, we are updating ctx in functions